provider/pdns.go

```diff
@@ -0,0 +1,369 @@
+// ref: https://doc.powerdns.com/authoritative/http-api/zone.html#get--servers-server_id-zones-zone_id
+func (c *PDNSAPIClient) ListZone(zoneID string) (pgo.Zone, *http.Response, error) {
+	zones, resp, err := c.client.ZonesApi.ListZone(c.authCtx, defaultServerID, zoneID)
+	if err != nil {
+		log.Warnf("Unable to list zone %s. %v", zoneID, err)
+	}
+	return zones, resp, err
+}
+
+// PatchZone : Method used to update the contents of a particular zone from PowerDNS
+// ref: https://doc.powerdns.com/authoritative/http-api/zone.html#patch--servers-server_id-zones-zone_id
+func (c *PDNSAPIClient) PatchZone(zoneID string, zoneStruct pgo.Zone) (*http.Response, error) {
+	resp, err := c.client.ZonesApi.PatchZone(c.authCtx, defaultServerID, zoneID, zoneStruct)
+	if err != nil {
+		log.Warnf("Unable to patch zone %s. %v", zoneStruct.Name, err)
+	}
+	return resp, err
+}
+
+// PDNSProvider is an implementation of the Provider interface for PowerDNS
+type PDNSProvider struct {
+	client PDNSAPIProvider
+}
+
+// NewPDNSProvider initializes a new PowerDNS based Provider.
+func NewPDNSProvider(server string, apikey string, domainFilter DomainFilter, dryRun bool) (*PDNSProvider, error) {
+
+	// Do some input validation
+
+	if apikey == "" {
+		return nil, errors.New("Missing API Key for PDNS. Specify using --pdns-api-key=")
+	}
+
+	// The default for when no --domain-filter is passed is [""], instead of [], so we check accordingly.
+	if len(domainFilter.filters) != 1 && domainFilter.filters[0] != "" {
+		return nil, errors.New("PDNS Provider does not support domain filter")
+	}
+	// We do not support dry running, exit safely instead of surprising the user
+	// TODO: Add Dry Run support
+	if dryRun {
+		return nil, errors.New("PDNS Provider does not currently support dry-run")
+	}
+
+	if server == "localhost" {
+		log.Warnf("PDNS Server is set to localhost, this may not be what you want. Specify using --pdns-server=")
+	}
+
+	cfg := pgo.NewConfiguration()
+	cfg.Host = server
+	cfg.BasePath = server + apiBase
+
+	provider := &PDNSProvider{
+		client: &PDNSAPIClient{
+			dryRun:  dryRun,
+			authCtx: context.WithValue(context.TODO(), pgo.ContextAPIKey, pgo.APIKey{Key: apikey}),
+			client:  pgo.NewAPIClient(cfg),
+		},
+	}
+
+	return provider, nil
+}
+
+func (p *PDNSProvider) convertRRSetToEndpoints(rr pgo.RrSet) (endpoints []*endpoint.Endpoint, _ error) {
+	endpoints = []*endpoint.Endpoint{}
+
+	for _, record := range rr.Records {
+		// If a record is "Disabled", it's not supposed to be "visible"
+		if !record.Disabled {
+			endpoints = append(endpoints, endpoint.NewEndpointWithTTL(rr.Name, record.Content, rr.Type_, endpoint.TTL(rr.Ttl)))
+		}
+	}
+
+	return endpoints, nil
+}
+
+// ConvertEndpointsToZones marshals endpoints into pdns compatible Zone structs
+func (p *PDNSProvider) ConvertEndpointsToZones(eps []*endpoint.Endpoint, changetype pdnsChangeType) (zonelist []pgo.Zone, _ error) {
+	/* eg of mastermap
```

Comment:
Is this a leftover for developing purpose? If not, I would move it to the comment of the method, otherwise just drop it.

Reply:
It's a stale comment from a previous implementation, I'll remove it.